### PR TITLE
softmax: add device check for xpu with half_to_float

### PR DESCRIPTION
**Pull request by weishi-deng**
We improved the kernel implementation for softmax with half input and float output for Intel GPUs( https://github.com/intel/torch-xpu-ops/pull/1516). The optimized kernel fused the data type casting and softmax computation. To apply the optimization, we need to allow `XPU` to be dispatched to `at::_softmax` with `half_to_float` being `true`, say `at::_softmax(input_, dim_, true)`.

Performance Improvement:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wdeng/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wdeng/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">

reduction | shape | PR improvement %
-- | -- | --
dim=0 | [64, 64] | 175%
dim=0 | [8192, 8192] | 160%
dim=0 | [64, 8192] | 104%
dim=0 | [8192, 64] | 163%
dim=0 | [1024, 1024] | 208%
dim=0 | [16384, 16384] | 156%
dim=1 | [64, 64] | 252%
dim=1 | [8192, 8192] | 151%
dim=1 | [64, 8192] | 224%
dim=1 | [8192, 64] | 214%
dim=1 | [1024, 1024] | 234%
dim=1 | [16384, 16384] | 151%

</body>

</html>

cc @gujinghui @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

**Comment by weishi-deng** [comment]
This pr is pending on the PR on torch-xpu-ops https://github.com/intel/torch-xpu-ops/pull/1516, please do not merge until the PR for third-party is merged.

**Review by guangyey — APPROVED** [review]

File: aten/src/ATen/native/SoftMax.cpp [hunk 1]
```diff
@@ -411,7 +411,7 @@ TORCH_IMPL_FUNC(log_softmax_backward_cpu_out) (
 Tensor softmax(const Tensor& input_, const int64_t dim_, std::optional<ScalarType> dtype) {
   auto result = [&]() {
     NoNamesGuard guard;
-    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+    if ((input_.is_cuda()||input_.is_xpu()) && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
```
Comment (guangyey) [inline aten/src/ATen/native/SoftMax.cpp:414]:
```suggestion
    if ((input_.is_cuda() || input_.is_xpu()) && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float) {
```

File: aten/src/ATen/native/SoftMax.cpp [hunk 2]
```diff
@@ -428,7 +428,7 @@ Tensor& softmax_out(
     std::optional<ScalarType> dtype,
     Tensor& output_) {
   Tensor output_temp;
-  if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half &&
+  if ((input_.is_cuda()||input_.is_xpu()) && input_.scalar_type() == ScalarType::Half &&
```
Comment (guangyey) [inline aten/src/ATen/native/SoftMax.cpp:431]:
```suggestion
  if ((input_.is_cuda() || input_.is_xpu()) && input_.scalar_type() == ScalarType::Half &&
```

File: aten/src/ATen/native/SoftMax.cpp [hunk 3]
```diff
@@ -467,7 +467,7 @@ Tensor special_softmax(const Tensor& input_, const int64_t dim_, std::optional<S
 Tensor log_softmax(const Tensor& input_, const int64_t dim_, std::optional<ScalarType> dtype) {
   auto result = [&]() {
     NoNamesGuard guard;
-    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+    if ((input_.is_cuda()||input_.is_xpu()) && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
```
Comment (guangyey) [inline aten/src/ATen/native/SoftMax.cpp:470]:
```suggestion
    if ((input_.is_cuda() || input_.is_xpu()) && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
```

File: aten/src/ATen/native/SoftMax.cpp [hunk 4]
```diff
@@ -484,7 +484,7 @@ Tensor& log_softmax_out(
     std::optional<ScalarType> dtype,
     Tensor& output_) {
   Tensor output_temp;
-  if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half &&
+  if ((input_.is_cuda()||input_.is_xpu()) && input_.scalar_type() == ScalarType::Half &&
```
Comment (guangyey) [inline aten/src/ATen/native/SoftMax.cpp:487]:
```suggestion
  if ((input_.is_cuda() || input_.is_xpu()) && input_.scalar_type() == ScalarType::Half &&
```

**Comment by weishi-deng** [comment]
Ready for review as PR https://github.com/intel/torch-xpu-ops/pull/1516 landed. Please help review again. @guangyey

**Comment by guangyey** [comment]
I think this PR depends on torch-xpu-ops commit pin update.
cc @xytintel

**Comment by guangyey** [comment]
Additionally, you need to add a UT to ensure half_to_float functionality works expectantly.

**Comment by weishi-deng** [comment]
> Additionally, you need to add a UT to ensure half_to_float functionality works expectantly.

We add a unit test in torch-xpu-ops.  Do we have to add one in pytorch?

**Comment by guangyey** [comment]
> We add a unit test in torch-xpu-ops. Do we have to add one in pytorch?

This PR hasn't landed yet, so the unit tests in torch-xpu-ops can't verify the half_to_float functionality. To ensure this functionality is properly tested and visible to reviewers, we should include a unit test within this PR.

File: test/xpu/test_softmax.py
Comment (EikanWang) [inline test/xpu/test_softmax.py:1]: @weishi-deng , The `softmax` is part of `op_db`. Let's reuse the `op_db` rather than an new dedicated test file.

File: aten/src/ATen/native/SoftMax.cpp [hunk 5]
```diff
@@ -411,7 +411,7 @@ TORCH_IMPL_FUNC(log_softmax_backward_cpu_out) (
 Tensor softmax(const Tensor& input_, const int64_t dim_, std::optional<ScalarType> dtype) {
   auto result = [&]() {
     NoNamesGuard guard;
-    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+    if ((input_.is_cuda() || input_.is_xpu()) && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float) {
```
Comment (EikanWang) [inline aten/src/ATen/native/SoftMax.cpp:414]: @weishi-deng , what's the impact? Could you elaborate on the motivation? Does this change lead to any performance improvement? Or does it intend to fix failed cases?

File: test/xpu/test_softmax.py
Comment (guangyey) [reply test/xpu/test_softmax.py:1]: I think this should be a syntax sugar only for `softmax` to support input is `torch.half` but the output is `torch.float`, this is not a common test on `op_db`, right?
Anyway, we should not add a new test file to test this feature, a better approach is moving this test to https://github.com/pytorch/pytorch/blob/main/test/test_xpu.py

File: test/xpu/test_softmax.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            for j in range(len(shape[i])):
+                dim = j - 1
+                x = torch.randn(shape[i]).to(input_type)
+                grad = torch.randn(shape[i]).to(output_type)
+                x_cpu = x.clone().requires_grad_()
+                y_cpu = F.softmax(x_cpu, dim, dtype=output_type)
+                y_cpu.backward(grad.clone())
+
+                x_xpu = x.clone().to(device).requires_grad_()
+                y_xpu = F.softmax(x_xpu, dim, dtype=output_type)
```
Comment (guangyey) [inline test/xpu/test_softmax.py:51]: Actually, this PR primarily aims to verify that y_xpu is of type `torch.float32`.
```suggestion
                y_xpu = F.softmax(x_xpu, dim, dtype=output_type)
                self.assertEqual(y_xpu.dtype, torch.float32)
```

**Comment by guangyey** [comment]
@weishi-deng Please resolve conflicts to rebase this PR.

File: test/test_xpu.py [hunk 7]
```diff
@@ -732,6 +732,38 @@ def test_get_arch_list(self):
         if not torch.xpu._is_compiled():
             self.assertEqual(len(torch.xpu.get_arch_list()), 0)
 
+@unittest.skipIf(not TEST_XPU, "XPU not available, skipping tests")
+class TestSoftmaxWithHalfToFloat(TestCase):
```
Comment (guangyey) [inline test/test_xpu.py:736]: I think it is either
- move `test_softmax_half_to_float` to `TestXpu` class, or
- rename `TestSoftmaxWithHalfToFloat` to a common name such as `TestXpuOp` and move all op-related cases such as `test_compare_cpu` and `test_non_standard_bool_values` here

File: test/test_xpu.py [hunk 8]
```diff
@@ -732,6 +682,88 @@ def test_get_arch_list(self):
...
+                y_cpu.backward(grad.clone())
+
+                x_xpu = x.clone().to(device).requires_grad_()
+                y_xpu = torch.nn.functional.softmax(x_xpu, dim, dtype=output_type)
+                self.assertEqual(y_xpu.dtype, torch.float32)
+                y_xpu.backward(grad.clone().to(device))
+                self.assertEqual(y_cpu, y_xpu.cpu())
+                self.assertEqual(x_cpu.grad, x_xpu.grad.cpu())
+
+    @onlyXPU
```
Comment (guangyey) [inline test/test_xpu.py:718]:
```suggestion
```

File: test/test_xpu.py [hunk 9]
```diff
@@ -732,6 +682,87 @@ def test_get_arch_list(self):
...
+            cpu_results = op(cpu_sample.input, *cpu_sample.args, **cpu_sample.kwargs)
+
+            xpu_results = sample.output_process_fn_grad(xpu_results)
+            cpu_results = cpu_sample.output_process_fn_grad(cpu_results)
+
+            # Lower tolerance because we are running this as a `@slowTest`
+            # Don't want the periodic tests to fail frequently
+            self.assertEqual(xpu_results, cpu_results, atol=1e-4, rtol=1e-4)
+
+    @onlyXPU
```
Comment (guangyey) [inline test/test_xpu.py:740]:
```suggestion
```

File: test/test_xpu.py [hunk 10]
```diff
@@ -732,6 +682,86 @@ def test_get_arch_list(self):
...
+            self.assertEqual(ret, x)
+            return ret
+
+        for sample in op.sample_inputs(device, dtype):
+            expect = op(sample.input, *sample.args, **sample.kwargs)
+
+            transformed = sample.transform(convert_boolean_tensors)
+            actual = op(transformed.input, *transformed.args, **transformed.kwargs)
+
+            self.assertEqual(expect, actual)
```
Comment (guangyey) [inline test/test_xpu.py:764]:
```suggestion
            self.assertEqual(expect, actual)


instantiate_device_type_tests(TestXpuOps, globals(), only_for="xpu", allow_xpu=True)

```

File: test/test_xpu.py [hunk 11]
```diff
@@ -732,6 +682,90 @@ def test_get_arch_list(self):
         if not torch.xpu._is_compiled():
             self.assertEqual(len(torch.xpu.get_arch_list()), 0)
 
+@unittest.skipIf(not TEST_XPU, "XPU not available, skipping tests")
+class TestXpuOps(TestCase):
+    def test_softmax_half_to_float(self):
```
Comment (guangyey) [inline test/test_xpu.py:687]:
```suggestion
    def test_softmax_half_to_float(self, device):
```

File: test/test_xpu.py [hunk 12]
```diff
@@ -732,6 +682,90 @@ def test_get_arch_list(self):
...
+            [7, 8],
+            [8192, 64],
+            [8192, 8192],
+            [7, 8, 512],
+            [7, 8, 11],
+            [16, 7, 8, 512],
+            [16, 7, 8, 512, 35],
+            [117, 7, 9, 513, 35],
+        ]
+        device = "xpu"
```
Comment (guangyey) [inline test/test_xpu.py:699]:
```suggestion
```

File: test/test_xpu.py [hunk 13]
```diff
@@ -732,6 +682,89 @@ def test_get_arch_list(self):
         if not torch.xpu._is_compiled():
             self.assertEqual(len(torch.xpu.get_arch_list()), 0)
 
+@unittest.skipIf(not TEST_XPU, "XPU not available, skipping tests")
+class TestXpuOps(TestCase):
+    def test_softmax_half_to_float(self, device):
```
Comment (guangyey) [inline test/test_xpu.py:687]:
```suggestion
    @dtypes(torch.float16)
    def test_softmax_half_to_float(self, device, dtype):
```

File: test/test_xpu.py [hunk 14]
```diff
@@ -732,6 +679,89 @@ def test_get_arch_list(self):
         if not torch.xpu._is_compiled():
             self.assertEqual(len(torch.xpu.get_arch_list()), 0)
 
+@unittest.skipIf(not TEST_XPU, "XPU not available, skipping tests")
```
Comment (guangyey) [inline test/test_xpu.py:682]:
```suggestion

@unittest.skipIf(not TEST_XPU, "XPU not available, skipping tests")
```

**Review by guangyey — APPROVED** [review]

File: aten/src/ATen/native/SoftMax.cpp (hunk 5 above)
Comment (weishi-deng) [reply aten/src/ATen/native/SoftMax.cpp:414]: > @weishi-deng , what's the impact? Could you elaborate on the motivation? Does this change lead to any performance improvement? Or does it intend to fix failed cases?

This pr makes the softmax op support half input and float output without casting. This functionality has been added in torch-xpu-ops, so we need to add the device option here to enable it. Otherwise, it needs to cast the input before we do softmax, as lines 417-418 say.

Comment (EikanWang) [reply aten/src/ATen/native/SoftMax.cpp:414]: @weishi-deng , as we discussed, it should not be a feature improvement. The change is a performance improvement as the torch-xpu-ops has provided an optimal implementation to fuse the data type cast and softmax. So, pls. share the performance improvement data in the PR description.

Comment (weishi-deng) [reply aten/src/ATen/native/SoftMax.cpp:414]: > @weishi-deng , as we discussed, it should not be a feature improvement. The change is a performance improvement as the torch-xpu-ops has provided an optimal implementation to fuse the data type cast and softmax. So, pls. share the performance improvement data in the PR description.

Hi @EikanWang, The performance data is updated in the PR description. Pls review and comment.

**Comment by EikanWang** [comment]
@weishi-deng , I suppose the `perf %(PR vs main)` should be higher is better, right? If so, do you know why the following two shapes have regression?

single ut with   softmax(shape, dim) | perf %(PR vs main)
-- | --
[8192, 64] , 0 | 76%
[1024, 1024], 0 | 71%

**Comment by EikanWang** [comment]
By the way, please check the test case failure.

**Comment by weishi-deng** [comment]
> @weishi-deng , I suppose the `perf %(PR vs main)` should be higher is better, right? If so, do you know why the following two shapes have regression?
> 
> single ut with softmax(shape, dim)	perf %(PR vs main)
> [8192, 64] , 0	76%
> [1024, 1024], 0	71%

These cases are cases with strided load and strided store(dim =0). In the kernel implementation, we set the vectorized length as 16Byte/sizeof(input scalar), so the vectorization length grows from float-4 to half-8 when we enable the input with half type. For the strided instances, it increases the memory load instructions with the stride per thread. As synced with @xytintel, we think the perf drop for these cases is expected.

**Comment by EikanWang** [comment]
> In the kernel implementation, we set the vectorized length as 16Byte/sizeof(input scalar), so the vectorization length grows from float-4 to half-8 when we enable the input with half type.

Have you evaluated the optimization heuristic rule on Arc B-series?

> For the strided instances, it increases the memory load instructions with the stride per thread.

Is there any data to support the conclusion?

> As synced with @xytintel, we think the perf drop for these cases is expected.

Frankly speaking, it should not be as expected. In general, we don't expect any performance regression. Let's collect more E2E performance data.

**Comment by weishi-deng** [comment]
The perf statistics are updated as the softmax implementation is optimized according to PRhttps://github.com/intel/torch-xpu-ops/pull/1781
